### ISSUE TITLE
[x86][houdini] UnsatisfiedLinkError when running on Android devices with x86 processors

### ISSUE THREAD

**Issue by krunalsshah** [facebook/react-native#12246]
# SoLoader$WrongAbiError: APK was built for a different platform

### Description
For app that supports only armeabi-v7a libraries on playstore (running on Intel Z2560 and other chip sets with dual ABI support) causes error  `java.lang.UnsatisfiedLinkError: dlopen failed: "/data/app-lib/com.app.android-4/libreactnativejni.so" has unexpected e_machine: 40`

Ideally [houdini](https://commonsware.com/blog/2013/11/21/libhoudini-what-it-means-for-developers.html) ARM translator should do the trick by allowing NDK binaries to run on x86 hardware. 

``` 
D/houdini: [12822] Loading library(version: 4.0.8.45720 RELEASE)... successfully.
[12822] Unsupported feature (ID:0x10600cae).
[12822] Open Native Library /data/app-lib/com.walmart.android-4/libreactnativejni.so failed.
```
The logs above clearly suggest that `libhoudini` failed to open`libreactnativejni.so`. I am aware by including `x86` support for react binaries the problem can be solved, however we currently aren't considering the option (including split apk option). Does react-native team have any other alternative solution?

The related Issue can be found [here](https://github.com/facebook/react-native/issues/5773)

### Reproduction
Please find the steps [HERE](https://github.com/facebook/react-native/issues/5773#issuecomment-255567298) 
In a nut shell , create a new React native project and remove "x86" from the abiFilters property in the build.gradle file.

### Additional Information

* React Native version: 0.40	
* Platform: Android, Tablets with Intel Z2560 and other chip sets with dual ABI support
* Operating System: Linux
* On running `adb shell getprop | grep ro.product.cpu` the device properties are rightly configured. 
```
[ro.product.cpu.abi2]: [armeabi-v7a]
[ro.product.cpu.abi]: [x86]
 ```
* On stepping into the code to check the abi values programmatically, the abi values are returning values identical to lib folder of apk , Please check the png file. Any idea?
```
        final String cpuAbi = Build.CPU_ABI; // value = armeabi-v7a
        final String cpuAbi2 = Build.CPU_ABI2; // value = armeabi
 ```
<img width="529" alt="apk" src="https://cloud.githubusercontent.com/assets/6138310/22672371/558e797c-ec88-11e6-9b5d-272aa33dbea2.png">

**Comment by matthewfang** [comment]
the same issue, i have this problem with RN 0.42. how can we solve it?

**Comment by yaxin3690** [comment]
是是是

**Comment by wj495175289** [comment]
the same issue, i have this problem with RN 0.43.4. how can we solve it?

**Comment by krunalsshah** [comment]
At this point the solution is to include `abiFilters` as below in `defaultConfig` of your applications `build.gradle`. 
```
ndk {
  abiFilters "armeabi-v7a", "x86"
}
```
Yes this will increase the size of the .apk file. If you want to reduce the apk file size you should go with [split APKs](https://developer.android.com/studio/build/configure-apk-splits.html) Please note universal .apk will not work as dual ABI support devices will still try to load `arm` .so files and still fail with `java.lang.UnsatisfiedLinkError: dlopen failed: "/data/app-lib/com.app.android-4/libreactnativejni.so" has unexpected e_machine: 40`

**Comment by yaxin3690** [comment]
thinks @krunalsshah

**Comment by lyahdav** [comment]
Just to be clear, this may be a roadblock for some apps to integrate React Native because there is some small percentage of Android devices that run x86 processors. We end up with a few non-ideal options:

1. Don't support x86 devices. This isn't feasible for some apps, as our native (Java) code runs fine on these devices and we don't want to drop support for those users.
1. Include `x86` in ABI filters (mentioned above [here](https://github.com/facebook/react-native/issues/12246#issuecomment-308749310)), which increases the binary size a lot (10 MB in our case), which is prohibitive.
1. Switch to a [Split APK](https://developer.android.com/studio/build/configure-apk-splits.html), which for some teams is a prohibitively large amount of work because of assumptions in our build and release tooling about app versions.

It's not clear to be why other libraries with native dependencies work fine with the [houdini](https://commonsware.com/blog/2013/11/21/libhoudini-what-it-means-for-developers.html) ARM translator but React Native's dependencies don't. It seems this is something that should be fixable.

**Comment by krunalsshah** [comment]
@lyahdav Precisely React Native `core` library needs to be fixed at the point where `houdini` fails to compile. This would avoid all of the above points mentioned above. Please let us know if you were able to find any alternative solution or if you happen to fix the `source`. Thanks.

**Comment by hramos** [comment]
Thanks for the additional details, @lyahdav. I'll try to find a point of contact that can help out here.

**Comment by hramos** [comment]
I've marked #5773 as a duplicate of this issue. If you're investigating this issue, I recommend reading through the comments in #5773 for additional details.

Specifically, @lyahdav has posted some findings [here](https://github.com/facebook/react-native/issues/5773#issuecomment-255567298), and [here](https://github.com/facebook/react-native/issues/5773#issuecomment-255568633)

**Comment by krunalsshah** [comment]
@hramos Thanks for taking a look at this. You'd want to take a look from the start as all the bullets captured are summarized in the previous steps. Also the original post summarizes the steps to reproduce etc. Please let us know if you have any questions.

**Comment by hramos** [comment]
@krunalsshah I've shared this issue internally and am looking for a point of contact who can help provide more context.

**Comment by zigbang-ethan** [comment]
Can anybody help?

**Comment by lyahdav** [comment]
I have made some progress on investigating this issue. I found that when I change the `APP_STL` value in [Application.mk](https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/jni/Application.mk) from `gnustl_shared` to `gnustl_static` I no longer get the "Unsupported feature" error when loading `libreactnativejni.so`. It still gives the "Unsupported feature" error on the other libraries, which I haven't had time investigate. I'm a bit concerned with changing `APP_ STL` to `static` because of the C++ [One Definition Rule](http://en.wikipedia.org/wiki/One_Definition_Rule), as mentioned [here](https://stackoverflow.com/a/25856207/62).

If anyone wants to investigate in parallel:
1. Clone the RN repo.
1. Follow Android prerequisites [here](https://facebook.github.io/react-native/docs/android-building-from-source.html).
1. Open RN project in Android Studio, remove "x86" from `abiFilters` in `RNTester/android/app/build.gradle`.
1. Run RNTester app in [BlueStacks emulator](https://www.bluestacks.com/) (which has Houdini).
1. Start making changes to files like Application.mk, Android.mk, etc. and see if it fixes anything.

**Comment by dryganets** [comment]
I had to solve it a few months back as we didn't have x86 libraries yet and need to support Intel devices anyway.

There are two problems:
1. There is a bug in the JSC library react-native using you need to rather turn off the optimizations or upgrade to the newer version of JSC to mitigate it.
2. On Android L devices there is a bug in libhoudini dlclose implementation. You need to remove the dlclose ASSERT to fix it. You need it only in case you are using Cxx modules though.

**Comment by qiyue5755** [comment]
@lyahdav I am sorry, I also have this problem with 0.51.0？Is there a solution to this problem?
